refactor(ai-builder): Add PromptBuilder and registry system for prompts (no-changelog)

OlegIvaniv opened the pull request:
## Summary

Introduces a PromptBuilder utility for type-safe prompt composition and a registry-based system for dynamically selecting node-specific guides and examples. This makes it easy to add best practices for new nodes by simply creating a guide file and registering it.

## Changes

- Add `PromptBuilder` class with fluent API for composing prompts with conditional sections, priority ordering, and LangChain cache_control support
- Add registry system that pattern-matches node types to automatically include relevant guides and examples
- Migrate all agent prompts (supervisor, discovery, builder, configurator, responder) to use PromptBuilder
- Simplify `parameter-updater/` directory structure:
  - Merge `node-types/` and `parameter-types/` into `guides/`
  - Flatten `examples/` (remove `basic/` and `advanced/` nesting)
  - Combine base prompts into single `parameter-updater.prompt.ts`
- Remove old `ParameterUpdatePromptBuilder` class and `prompt-config.ts`
- Update README with current structure and guide for adding new node guides

## Related Linear tickets, Github issues, and Community forum posts

https://linear.app/n8n/issue/AI-1746

## Review / Merge checklist

- [x] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md))
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [x] Tests included.
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

OlegIvaniv committed:
feat(ai-builder): Add PromptBuilder for type-safe prompt composition

Introduces a new fluent builder API for constructing prompts with:
- `section()` for always-included sections
- `sectionIf()` for conditional sections with lazy evaluation
- `examples()` and `examplesIf()` for typed few-shot examples
- XML tag formatting by default (configurable to Markdown)
- Priority-based section ordering
- `buildAsMessageBlocks()` for LangChain integration with cache_control
- `estimateTokens()` for token budgeting
- `merge()` for composing multiple builders

Includes 57 unit tests covering all functionality.

OlegIvaniv committed:
refactor(ai-builder): Migrate simple agent prompts to PromptBuilder

Migrate supervisor, responder, and configurator prompts to use the new
PromptBuilder API with XML-tagged sections for better structure.

OlegIvaniv committed:
refactor(ai-builder): Migrate builder and discovery prompts to PromptBuilder

Migrate medium complexity prompts:
- builder.prompt.ts: 16 sections with priority ordering
- discovery.prompt.ts: Conditional sections via helper functions

OlegIvaniv committed:
refactor(ai-builder): Migrate ParameterUpdatePromptBuilder to use PromptBuilder

Refactor the complex ParameterUpdatePromptBuilder to use the new
PromptBuilder internally with sectionIf() for conditional node-type
specific guides and priority ordering for section sequencing.

OlegIvaniv committed:
refactor(ai-builder): Add XML tags to legacy agent prompt constants

Add explicit XML tags to RESPONSE_PATTERNS and PREVIOUS_SUMMARY
constants for consistency with other prompts. The legacy agent prompt
retains direct LangChain ChatPromptTemplate construction due to strict
type requirements for cache_control.

OlegIvaniv committed:
refactor(ai-builder): Replace ParameterUpdatePromptBuilder class with functions

Replace the static class pattern with simple exported functions:
- `buildParameterUpdatePrompt(context)` - builds the dynamic prompt
- `hasResourceLocatorParameters(nodeDefinition)` - utility for detection
- `estimatePromptTokens(prompt)` - token estimation

This simplifies the API and aligns with the functional approach used
by the new PromptBuilder utility.

OlegIvaniv committed:
feat(ai-builder): Add registry system for node-type prompts

Introduce a registration-based system for node-type specific guides
and examples. Features:

- Pattern matching with wildcards (*Tool, n8n-nodes-base.*, etc.)
- Priority-based ordering (lower = earlier)
- Conditional inclusion via predicate functions
- Case-insensitive matching

This enables auto-discovery of prompts based on node type.

OlegIvaniv committed:
refactor(ai-builder): Convert node-type guides to use registerGuide()

Update all node-type guide files to self-register with the registry:
- set-node.ts: patterns ['n8n-nodes-base.set', '.set']
- if-node.ts: patterns ['n8n-nodes-base.if', '.if']
- switch-node.ts: patterns ['n8n-nodes-base.switch', '.switch']
- http-request.ts: patterns for HTTP and webhook nodes
- tool-nodes.ts: pattern '*Tool' for all tool nodes

OlegIvaniv committed:
refactor(ai-builder): Convert parameter-type guides with conditions

Update parameter-type guides to self-register with condition functions:
- resource-locator.ts: shown when hasResourceLocatorParams is true
- system-message.ts: shown when node has system message parameters
- text-fields.ts: shown when node has string text fields

Each guide uses pattern '*' with a condition predicate for context-aware
inclusion based on node definition analysis.

OlegIvaniv committed:
refactor(ai-builder): Convert examples to use registerExamples()

Update all example files to self-register with the registry:
- set-node-examples.ts: priority 20, matches Set nodes
- if-node-examples.ts: priority 20, matches IF nodes
- switch-node-examples.ts: priority 20, matches Switch nodes
- tool-node-examples.ts: priority 20, matches *Tool pattern
- resource-locator-examples.ts: priority 30, conditional on hasResourceLocatorParams
- simple-updates.ts: priority 100 (fallback for unmatched nodes)

OlegIvaniv committed:
refactor(ai-builder): Use registry in parameter-updater chain

Update the parameter-updater chain to:
- Use PromptBuilder API directly instead of buildParameterUpdatePrompt
- Query guides and examples from registry via getMatchingGuides/getMatchingExamples
- Dynamically build prompt based on matched content

Also update index exports to expose registry functions and types
while maintaining backwards compatibility with existing exports.

OlegIvaniv committed:
refactor(ai-builder): centralize guide/example registration

- Replace self-registration pattern with centralized imports in registry.ts
- Add GuideConfig and ExamplesConfig interfaces for predictable exports
- Remove registerGuide/registerExamples functions and side effects

OlegIvaniv committed:
refactor(ai-builder): simplify guide/example exports

- Export NodeTypeGuide/NodeTypeExamples directly with content included
- Remove intermediate GuideConfig/ExamplesConfig types
- Simplify registry to just import and use guides/examples directly

OlegIvaniv committed:
Cleanup old chain prompt builder

Signed-off-by: Oleg Ivaniv <me@olegivaniv.com>

OlegIvaniv committed:
feat(ai-builder): add withExamples() and default formatter to PromptBuilder

- Add withExamples() method to append examples to the last section
- Add default formatter for examples() and examplesIf() that handles
  strings and objects with a `content` property
- withExamples() uses XML tags by default, markdown in markdown mode
- Change examples separator from \n to \n\n for better readability

OlegIvaniv committed:
refactor(ai-builder): integrate SUPERVISOR_PROMPT_SUFFIX into builder

- Move suffix content into buildSupervisorPrompt() as an instruction section
- Remove separate SUPERVISOR_PROMPT_SUFFIX export
- Simplify supervisor agent to use buildSupervisorPrompt() directly

OlegIvaniv committed:
refactor(ai-builder): use examplesIf() in parameter-updater chain

- Replace manual examples building with examplesIf() using default formatter
- Examples are now their own section instead of nested in common_patterns
- Fix IF node guide test to match actual content

OlegIvaniv committed:
refactor(ai-builder): simplify parameter-updater directory structure

- Combine base prompts into single parameter-updater.prompt.ts
- Merge node-types/ and parameter-types/ into guides/
- Flatten examples/ directory (remove basic/advanced subdirs)
- Keep registry at root level as central module
- Update all imports to use new structure

OlegIvaniv committed:
docs(ai-builder): update prompts README with current structure

- Update directory structure to reflect simplified parameter-updater
- Add PromptBuilder utility section with key features
- Document registry-based parameter updater system
- Add guide for adding new node guides/examples
- Remove fragile import statements, focus on concepts

OlegIvaniv committed:
refactor(ai-builder): remove unused priority configuration from PromptBuilder

Priority-based section ordering was not actually used - sections are now
output in insertion order which is simpler and sufficient for all use cases.

- Remove priority from SectionOptions and SectionEntry types
- Remove priority sorting from PromptBuilder and registry
- Remove priority field from all guides and examples
- Remove priority-related tests
- Update README documentation

OlegIvaniv committed:
fix(ai-builder): replace type casts with proper type guards

Address PR review feedback:
- system-message.ts: use type guard instead of `as` cast for opt.name
- prompt-builder.ts: remove unnecessary cast after `in` check

OlegIvaniv committed:
Cleanup

Signed-off-by: Oleg Ivaniv <me@olegivaniv.com>

mike12345567 commented on packages/@n8n/ai-workflow-builder.ee/src/prompts/chains/parameter-updater/guides/system-message.ts:
```diff
@@ -1,4 +1,42 @@
+
+	return nodeDefinition.properties.some((prop) => {
+		// Pattern 1 & 2: options.systemMessage (AI Agent) or options.system (Anthropic)
+		if (prop.name === 'options' && prop.type === 'collection') {
```
Nit: double if statement is a little less readable here - I'd love the pattern 3 check first to demonstrate always true scenarios up front, then pattern 1 & 2 collapse into a single if, like:
```ts
	// Pattern 1: messages parameter with role support
	if (
		prop.name === 'messages' &&
		(prop.type === 'fixedCollection' || prop.type === 'collection')
	) {
		return true;
	}

	// Pattern 2 & 3: options.systemMessage / options.system
	if (prop.name === 'options' && prop.type === 'collection' && Array.isArray(prop.options)) {
		return prop.options.some(
			(opt) =>
				typeof opt === 'object' &&
				opt !== null &&
				'name' in opt &&
				(opt.name === 'systemMessage' || opt.name === 'system'),
		);
	}

    return false;
```

mike12345567 reviewed (dismissed):
Thanks for this - really love the implementation, love a chaining builder! Can't wait to get this in, specifically the new mechanism for providing examples/guides for nodes/groups of nodes is great, should make it easier to make slight adjustments when we see they are needed.

Left one minor nit, feel free to ignore.

OlegIvaniv committed:
fix(ai-builder): add Vector Store retrieve-as-tool mode for AI Agent

- Add connection parameter rule for Vector Store when used with AI Agent
- Clarify mode description to include AI Agent use case

OlegIvaniv committed:
refactor(ai-builder): improve readability of system message detection

- Reorder pattern checks to show simple "always true" case first
- Collapse nested if statements into single condition

OlegIvaniv committed:
Merge remote-tracking branch 'origin/master' into ai-1746-feature-make-it-easy-to-add-best-practices-for-new-nodes

# Conflicts:
#	packages/@n8n/ai-workflow-builder.ee/src/prompts/agents/builder.prompt.ts
#	packages/@n8n/ai-workflow-builder.ee/src/prompts/agents/configurator.prompt.ts
#	packages/@n8n/ai-workflow-builder.ee/src/prompts/chains/parameter-updater/guides/gmail.ts
#	packages/@n8n/ai-workflow-builder.ee/src/prompts/chains/parameter-updater/guides/http-request.ts
#	packages/@n8n/ai-workflow-builder.ee/src/prompts/chains/parameter-updater/index.ts
#	packages/@n8n/ai-workflow-builder.ee/src/prompts/chains/parameter-updater/prompt-builder.ts
#	packages/@n8n/ai-workflow-builder.ee/src/prompts/chains/parameter-updater/prompt-config.ts

mike12345567 reviewed (approved):
LGTM - thanks for addressing comment :)
